File: 3437-MaximumTotalDamageWithSpellCasting/3437-MaximumTotalDamageWithSpellCasting.py
# Last updated: 1/12/2026, 3:13:28 PM
class Solution:
    def maximumTotalDamage(self, power: List[int]) -> int:
        ct = Counter(power)
        ct = sorted(ct.items())
        dp = [0 for i in range(len(ct))]
        dp[0] = ct[0][0]*ct[0][1]
        for i in range(1,len(ct)):
            cur = dp[i-1]
            alt = ct[i][0]*ct[i][1]
            j = i-1
            while j>=0 and ct[j][0]+2>=ct[i][0]:
                j-=1
            if j>=0:
                alt = dp[j]+ct[i][0]*ct[i][1]
            dp[i] = max(cur, alt)
        print(ct)
        return dp[-1]

File: 3437-MaximumTotalDamageWithSpellCasting/3437-MaximumTotalDamageWithSpellCasting.cpp
// Last updated: 1/12/2026, 3:13:30 PM
class Solution {
public:
    long long maximumTotalDamage(vector<int>& power) {
        map<int, int> cnt;
        for(int p : power) {
            cnt[p]++;
        }

        map<int, int64_t> dp;
        dp[0] = 0;
        dp[1] = cnt[1];

        for(auto [p, pCnt] : cnt) {
            if(p == 1) continue;
            int64_t maxDamage = 0;

            auto dontTakeIt = dp.lower_bound(p);
            if(dontTakeIt != dp.begin()) {
                dontTakeIt = prev(dontTakeIt);
                maxDamage = max(maxDamage, dontTakeIt->second);
            }

            auto takeIt = dp.lower_bound(p - 2);
            int64_t takeItDamage = 1LL * pCnt * p;
            if(takeIt != dp.begin()) {
                takeIt = prev(takeIt);
                takeItDamage += takeIt->second;
            }
            maxDamage = max(maxDamage, takeItDamage);

            dp[p] = maxDamage;

            cout << maxDamage << "\n";
        }

        return prev(dp.end())->second;
    }
};